File: 3244. Shortest Distance After Road Addition Queries II.java
class Solution {
    public int[] shortestDistanceAfterQueries(int n, int[][] queries) {
        Map<Integer, Integer> graph = new HashMap<>();
        for (int i = 0; i < n-1; i++) {
            graph.put(i, i+1);
        }
        int[] answer = new int[queries.length];
        for (int i = 0; i < answer.length; i++) {
            int from = queries[i][0];
            int to = queries[i][1];
            if (!graph.containsKey(from) || graph.get(from) >= to) {
                answer[i] = graph.size();
                continue;
            }
            int j = graph.get(from);
            while (j < to) {
                j = graph.remove(j);
            }
            graph.put(from, to);
            answer[i] = graph.size();
        }
        return answer;
    }
}

File: go/3244. Shortest Distance After Road Addition Queries II.go
func shortestDistanceAfterQueries(n int, queries [][]int) []int {
	graph := make(map[int]int)
	for i := 0; i < n-1; i++ {
		graph[i] = i+1
	}
	answer := make([]int, len(queries))
	for i, q := range queries {
		from, to := q[0], q[1]
		if _, ok := graph[from]; !ok || graph[from] >= to {
			answer[i] = len(graph)
			continue
		}
		j := graph[from]
		for j < to {
			if next, ok := graph[j]; ok {
				delete(graph, j)
				j = next
			} else {
				break
			}
		}
        graph[from] = to
		answer[i] = len(graph)
	}
	return answer
}